README.md
id-gen
======

[![Haskell Programming Language](https://img.shields.io/badge/language-Haskell-blue.svg)](http://www.haskell.org)
[![BSD3 License](http://img.shields.io/badge/license-BSD3-brightgreen.svg)](https://tldrlegal.com/license/bsd-3-clause-license-%28revised%29)


Description
-----------

Framework for generating pseudo-random IDs.

Basic idea of this library is to use simple monotonically increasing sequence,
and turn it into sequence of seemingly random IDs. This is done by composing
multiple invertible functions, which as a whole is also invertible.


ChangeLog.md
# ChangeLog / ReleaseNotes


## Version 0.1.0.0

- Initial release.


src/Data/Id/Type.hs
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE TypeFamilies #-}
module Data.Id.Type
    ( AnId(..)
    , Id32(..)
    , Id64(..)
    , Id128(..)

    -- * IdOffset
    , IdOffset(..)
    , IdOffset32(..)
    , IdOffset64(..)
    , IdOffset128(..)

    -- * Utilities
    , genericShowsTaggedId
    )
  where

import Prelude (Bounded, Enum, fromIntegral, maxBound, succ)

import Data.Bits (FiniteBits, finiteBitSize)
import Data.Bool (not, otherwise)
import Data.Coerce (Coercible, coerce)
import Data.Eq (Eq, (==))
import Data.Function (($), (.))
import qualified Data.List as List (null)
import Data.Maybe (Maybe(Just, Nothing))
import Data.Ord (Ord, (>))
import Data.Typeable
    ( Typeable
    , showsTypeRep
    , tyConName
    , typeOf
    , typeRep
    , typeRepArgs
    , typeRepTyCon
    )
import Data.Word (Word16, Word32, Word64)
import Text.Show
    ( Show(showsPrec)
    , ShowS
    , showChar
    , showParen
    , shows
    , showString
    )

import Data.Default (Default(def))
import Data.LargeWord (Word128)


-- {{{ Id ---------------------------------------------------------------------

-- {{{ AnId -------------------------------------------------------------------

class
    ( Bounded a
    , Eq a
    , Ord a
    , Typeable a
    , IdOffset (Offset a)
    , IdOffsetNum (Offset a) ~ IdNum a
    ) => AnId a
  where
    type IdNum a :: *
    type Offset a :: *
    type IdTag a :: *

    bitSize :: a -> Word16
    default bitSize :: FiniteBits (IdNum a) => a -> Word16
    bitSize i = withIdAsNum i $ fromIntegral . finiteBitSize

    withIdAsNum :: a -> (IdNum a -> r) -> r
    default withIdAsNum :: Coercible a (IdNum a) => a -> (IdNum a -> r) -> r
    withIdAsNum i f = coerce f i

genericShowsTaggedId
    ::  ( Typeable t
        , AnId (a t)
        , Show (IdNum (a t))
        )
    => a t
    -> ShowS
genericShowsTaggedId i =
    showsType . showString " @" . showsTag . showChar ' ' . withIdAsNum i shows
      where
        showsType = showString . tyConName . typeRepTyCon $ typeOf i

        showsTag =
            showParen (not . List.null $ typeRepArgs rep) $ showsTypeRep rep
          where
            rep = typeRep i

-- }}} AnId -------------------------------------------------------------------

-- {{{ Id{32,64,128} ----------------------------------------------------------

-- | 32 bit ID space.
newtype Id32 t = Id32 Word32
  deriving (Bounded, Eq, Ord, Typeable)

-- | 64 bit ID space.
newtype Id64 t = Id64 Word64
  deriving (Bounded, Eq, Ord, Typeable)

-- | 128 bit ID space.
newtype Id128 t = Id128 Word128
  deriving (Bounded, Eq, Ord, Typeable)

instance Typeable t => Show (Id32 t) where
    showsPrec d = showParen (d > 10) . genericShowsTaggedId

instance Typeable t => Show (Id64 t) where
    showsPrec d = showParen (d > 10) . genericShowsTaggedId

instance Typeable t => Show (Id128 t) where
    showsPrec d = showParen (d > 10) . genericShowsTaggedId

instance Typeable t => AnId (Id32 t) where
    type IdNum (Id32 t) = Word32
    type Offset (Id32 t) = IdOffset32 t
    type IdTag (Id32 t) = t

instance Typeable t => AnId (Id64 t) where
    type IdNum (Id64 t) = Word64
    type Offset (Id64 t) = IdOffset64 t
    type IdTag (Id64 t) = t

instance Typeable t => AnId (Id128 t) where
    type IdNum (Id128 t) = Word128
    type Offset (Id128 t) = IdOffset128 t
    type IdTag (Id128 t) = t

-- }}} Id{32,64,128} ----------------------------------------------------------

-- }}} Id ---------------------------------------------------------------------

-- {{{ Offset -----------------------------------------------------------------

-- {{{ IdOffset ---------------------------------------------------------------

class (Bounded a, Default a, Eq a, Ord a, Typeable a) => IdOffset a where
    type IdOffsetNum a :: *

    next :: a -> Maybe a
    default next
        :: (Enum (IdOffsetNum a), Coercible a (IdOffsetNum a)) => a -> Maybe a
    next offset
      | offset == maxBound = Nothing
      | otherwise          = Just (coerce (withIdOffsetAsNum offset succ))

    withIdOffsetAsNum :: a -> (IdOffsetNum a -> r) -> r
    default withIdOffsetAsNum
        :: Coercible a (IdOffsetNum a) => a -> (IdOffsetNum a -> r)-> r
    withIdOffsetAsNum i f = f (coerce i)

-- }}} IdOffset ---------------------------------------------------------------

-- {{{ Offset{32,64,128} ------------------------------------------------------

newtype IdOffset32 t = IdOffset32 Word32
  deriving (Bounded, Eq, Ord, Show, Typeable)

newtype IdOffset64 t = IdOffset64 Word64
  deriving (Bounded, Eq, Ord, Show, Typeable)

newtype IdOffset128 t = IdOffset128 Word128
  deriving (Bounded, Eq, Ord, Show, Typeable)

instance Default (IdOffset32 t) where
    def = IdOffset32 0

instance Default (IdOffset64 t) where
    def = IdOffset64 0

instance Default (IdOffset128 t) where
    def = IdOffset128 0

instance Typeable t => IdOffset (IdOffset32 t) where
    type IdOffsetNum (IdOffset32 t) = Word32

instance Typeable t => IdOffset (IdOffset64 t) where
    type IdOffsetNum (IdOffset64 t) = Word64

instance Typeable t => IdOffset (IdOffset128 t) where
    type IdOffsetNum (IdOffset128 t) = Word128

-- }}} Offset{32,64,128} ------------------------------------------------------

-- }}} Offset -----------------------------------------------------------------


src/Data/Id/Utils.hs
{-# LANGUAGE NoImplicitPrelude #-}
module Data.Id.Utils
    (
    -- * Word32
      toOctetsFromWord32
    , fromOctetsToWord32

    -- * Word64
    , toOctetsFromWord64
    , fromOctetsToWord64
    )
  where

import Prelude (fromIntegral)

import Data.Word (Word32, Word64, Word8)
import Data.Bits ((.&.), (.|.), shiftL, shiftR)


toOctetsFromWord32 :: Word32 -> (Word8, Word8, Word8, Word8)
toOctetsFromWord32 x =
    ( fromIntegral ((x .&. 0xFF000000) `shiftR` 24)
    , fromIntegral ((x .&. 0x00FF0000) `shiftR` 16)
    , fromIntegral ((x .&. 0x0000FF00) `shiftR` 8 )
    , fromIntegral ( x .&. 0x000000FF             )
    )
{-# INLINE toOctetsFromWord32 #-}

toOctetsFromWord64
    :: Word64
    -> (Word8, Word8, Word8, Word8, Word8, Word8, Word8, Word8)
toOctetsFromWord64 x = toOctetsFromWord64'
    (fromIntegral (x .&. 0xFFFFFFFF00000000) `shiftR` 32)
    (fromIntegral (x .&. 0x00000000FFFFFFFF)            )
  where
    toOctetsFromWord64' w1 w2 =
        joinTuples (toOctetsFromWord32 w1) (toOctetsFromWord32 w2)

    joinTuples (o1, o2, o3, o4) (o5, o6, o7, o8) =
        (o1, o2, o3, o4, o5, o6, o7, o8)
{-# INLINE toOctetsFromWord64 #-}

fromOctetsToWord32 :: (Word8, Word8, Word8, Word8) -> Word32
fromOctetsToWord32 (o1, o2, o3, o4) =
    (fromIntegral o1 `shiftL` 24) .|. (fromIntegral o2 `shiftL` 16)
    .|. (fromIntegral o3 `shiftL` 8) .|. fromIntegral o4
{-# INLINE fromOctetsToWord32 #-}

fromOctetsToWord64
    :: (Word8, Word8, Word8, Word8, Word8, Word8, Word8, Word8)
    -> Word64
fromOctetsToWord64 (o1, o2, o3, o4, o5, o6, o7, o8) =
    fromOctetsToWord64' (o1, o2, o3, o4) (o5, o6, o7, o8)
  where
    fromOctetsToWord64' hi lo =
        fromIntegral (fromOctetsToWord32 hi `shiftL` 32)
        .|. fromIntegral (fromOctetsToWord32 lo)
{-# INLINE fromOctetsToWord64 #-}


src/Data/Id/Base16.hs
{-# LANGUAGE NoImplicitPrelude #-}
module Data.Id.Base16
    ( IdToBase16(..)
    , toBase16
    , toBase16'
    , IdFromBase16(..)
    )
  where

import Prelude (error)

import Data.Bool (not, otherwise)
import Data.Maybe (Maybe(Just, Nothing))
import Data.Ord ((<))
import Data.Typeable (Typeable)

import qualified Data.ByteString.Base16.Lazy as Base16 (decode)
import Data.ByteString.Builder (Builder)
import Data.ByteString.Lazy (ByteString)
import qualified Data.ByteString.Lazy as ByteString
    ( length
    , null
    , splitAt
    , unpack
    )
import qualified Data.ByteString.Builder as Builder
    ( toLazyByteString
    , word32HexFixed
    , word64HexFixed
    )

import Data.Id.Type (AnId, Id32(Id32), Id64(Id64), IdTag, withIdAsNum)
import Data.Id.Utils (fromOctetsToWord32, fromOctetsToWord64)


class AnId a => IdToBase16 a where
    {-# MINIMAL buildBase16' #-}

    buildBase16 :: a -> proxy (IdTag a) -> Builder
    buildBase16 a _proxy = buildBase16' a

    buildBase16' :: a -> Builder

instance Typeable t => IdToBase16 (Id32 t) where
    buildBase16' = (`withIdAsNum` Builder.word32HexFixed)

instance Typeable t => IdToBase16 (Id64 t) where
    buildBase16' = (`withIdAsNum` Builder.word64HexFixed)

toBase16 :: IdToBase16 a => a -> proxy (IdTag a) -> ByteString
toBase16 a p = Builder.toLazyByteString (buildBase16 a p)

toBase16' :: IdToBase16 a => a -> ByteString
toBase16' a = Builder.toLazyByteString (buildBase16' a)

-- TODO: instance IdToBase16 (Id128 t)

class AnId a => IdFromBase16 a where
    parseBase16 :: proxy (IdTag a) -> ByteString -> (Maybe a, ByteString)

instance Typeable t => IdFromBase16 (Id32 t) where
    parseBase16 _proxy bs
      | ByteString.length input < 8 = (Nothing, bs)
      | not (ByteString.null rest')  = (Nothing, bs)
      | otherwise = case ByteString.unpack input' of
            -- We now know that input' is exactly 4 octets long, since input is
            -- exactly 8 hexadecimal digits long. However, this only works only
            -- when the input was fully decoded, which is the reason why we are
            -- checking for empty rest'.
            [o1, o2, o3, o4] ->
                (Just (Id32 (fromOctetsToWord32 (o1, o2, o3, o4))), rest)
            _ -> error "parseBase16 @(Id32 t): Imposible case."
      where
        (input, rest) = ByteString.splitAt 8 bs
        (input', rest') = Base16.decode input

instance Typeable t => IdFromBase16 (Id64 t) where
    parseBase16 _proxy bs
      | ByteString.length input < 16 = (Nothing, bs)
      | not (ByteString.null rest')  = (Nothing, bs)
      | otherwise = case ByteString.unpack input' of
            -- We now know that input' is exactly 8 octets long, since input is
            -- exactly 16 hexadecimal digits long. However, this only works
            -- only when the input was fully decoded, which is the reason why
            -- we are checking for empty rest'.
            [o1, o2, o3, o4, o5, o6, o7, o8] ->
                let n = fromOctetsToWord64 (o1, o2, o3, o4, o5, o6, o7, o8)
                in (Just (Id64 n), rest)
            _ -> error "parseBase16 @(Id64 t): Imposible case."
      where
        (input, rest) = ByteString.splitAt 16 bs
        (input', rest') = Base16.decode input


src/Data/Id/Gen.hs
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE FunctionalDependencies #-}
-- | Generate 'Id32', 'Id64' and 'Id128' values.
module Data.Id.Gen
    (
    -- | We want to minimize the amount of information that has to be stored,
    -- but we also want to somewhat randomize the values of IDs.

    -- * GenId
      GenId(..)
    , genNextId

    -- * Quadratic Residue
    --
    -- $quadraticResidue

    -- ** Quadratic Residue for Word32
    , largestPrime32bit
    , quadraticResidue32bit

    -- ** Quadratic Residue for Word64
    , largestPrime64bit
    , quadraticResidue64bit

    -- ** Quadratic Residue for Word128
    , largestPrime128bit
    , quadraticResidue128bit

    -- * Bit Operations
    --
    -- Invertible functions that operate on bits of a word.
    , swapLoAndHiHalf
    , swapEveryOtherBit
    )
  where

import Prelude (Integral, (*), (-), fromIntegral, quot, rem)

import Data.Bits (FiniteBits, (.|.), bit, finiteBitSize, xor)
import qualified Data.Bits as Bits (rotate)
import Data.Bool (otherwise)
import Data.Foldable (foldr)
import Data.Function (($), (.), on)
import Data.Functor ((<$>))
import Data.Maybe (Maybe)
import Data.Ord (Ord((<=), (>=)))
import Data.Proxy (Proxy(Proxy))
import Data.Word (Word32, Word64)

import Data.LargeWord (Word128, Word160, Word96)

import Data.Id.Type (AnId(Offset), IdOffset(next))


-- {{{ GenId ------------------------------------------------------------------

-- | Class that describes how IDs in an ID space 'Id' are generated.
class AnId a => GenId r a | r -> a where
    -- | Generate another non-repeating ID from ID space. Value @r :: *@ is
    -- modified to ensure that the IDs do not repeate.
    --
    -- ID space can be limited, and 'Nothing' is returned when the space is
    -- exhausted. This is especially required if some kind of sharding is in
    -- place.
    genId :: r -> Maybe (a, r)

-- | Generate ID by using from least unused 'Offset'.
genNextId :: AnId a => (Offset a -> a) -> Offset a -> Maybe (a, Offset a)
genNextId f offset = (f offset,) <$> next offset

-- }}} GenId ------------------------------------------------------------------

-- {{{ Quadratic Residue ------------------------------------------------------

-- | Largest prime number that fits in to 'Word32' and is congruent to 3 modulo
-- 4.
--
-- @
-- 'largestPrime32bit' = 2^32 - 5 = 4294967291
-- @
--
-- Source:
-- <https://primes.utm.edu/lists/2small/0bit.html Primes just less than a power of two 8 to 100 bits>
largestPrime32bit :: Word32
largestPrime32bit = 4294967291

-- | Largest prime number that fits in to 'Word64' and is congruent to 3 modulo
-- 4.
--
-- @
-- 'largestPrime64bit' = 2^64 - 189 = 18446744073709551427
-- @
--
-- Source:
-- <https://primes.utm.edu/lists/2small/0bit.html Primes just less than a power of two 8 to 100 bits>
largestPrime64bit :: Word64
largestPrime64bit = 18446744073709551427

-- | Largest prime number that fits in to 'Word128' and is congruent to 3
-- modulo 4.
--
-- @
-- 'largestPrime128bit' = 2^128 - 173 = 340282366920938463463374607431768211283
-- @
--
-- Source:
-- <https://primes.utm.edu/lists/2small/100bit.html Primes just less than a power of two 101 to 200 bits>
largestPrime128bit :: Word128
largestPrime128bit = 340282366920938463463374607431768211283

-- | Generic implementation for quadratic residues, but it works correctly only
-- for non-negative integers. This function doesn't check if the input is
-- non-negative.
quadraticResidue :: Integral i => i -> i -> i
quadraticResidue p n
  | n >= p    = n   -- Out of range.
  | otherwise = if n <= (p `quot` 2) then q else p - q
      where
        q = (n * n) `rem` p
        -- Using quot and rem is faster than div and mod. It gives the same
        -- results for positive integers.
{-# INLINE quadraticResidue #-}

-- | Intermediate values used internally by 'quadraticResidue' may not fit in
-- to the type we want as a result.
convert :: (Integral a, Integral b) => proxy b -> (b -> b -> b) -> a -> a -> a
convert _ f x y = fromIntegral $ (f `on` fromIntegral) x y
{-# INLINE convert #-}

-- | Quadratic residue for 32b unsigned integers that uses 'largestPrime32bit'.
quadraticResidue32bit :: Word32 -> Word32
quadraticResidue32bit =
    convert (Proxy :: Proxy Word64) quadraticResidue largestPrime32bit
    -- Maximum value used internally is n^2 for n = largestPrime32bit - 1,
    -- which doesn't fit in to 32 bits, but it does in to 64 bits.
{-# INLINE quadraticResidue32bit #-}

-- | Quadratic residue for 64b unsigned integers that uses 'largestPrime64bit'.
quadraticResidue64bit :: Word64 -> Word64
quadraticResidue64bit =
    convert (Proxy :: Proxy Word96) quadraticResidue largestPrime64bit
    -- Maximum value used internally is n^2 for n = largestPrime64bit - 1,
    -- which doesn't fit in to 64 bits, but it does in to 96 bits.
{-# INLINE quadraticResidue64bit #-}

-- | Quadratic residue for 128b unsigned integers that uses
-- 'largestPrime128bit'.
quadraticResidue128bit :: Word128 -> Word128
quadraticResidue128bit =
    convert (Proxy :: Proxy Word160) quadraticResidue largestPrime128bit
    -- Maximum value used internally is n^2 for n = largestPrime128bit - 1,
    -- which doesn't fit in to 128 bits, but it does in to 160 bits.
{-# INLINE quadraticResidue128bit #-}

-- $quadraticResidue
--
-- Quadratic residue @q@ is obtained using:
--
-- @
-- x² ≡ q (mod p)
-- @
--
-- There are @(p + 1)/2@ residues (including 0) and @(p − 1)/2@ nonresidues. In
-- this case, it is customary to consider 0 as a special case and work within
-- the multiplicative group of nonzero elements of the field Z/pZ. (In other
-- words, every congruence class except zero modulo p has a multiplicative
-- inverse. This is not true for composite moduli.)
--
-- If @p ≡ 3 (mod 4)@ the negative of a residue modulo p is a nonresidue and
-- the negative of a nonresidue is a residue.
--
-- Algorithm used in here is based on the above principles and can be restated
-- as follows.
--
-- Quadratic residue @q@ is q = n² `mod` p, and it is unique as long as
-- @2n < p+1 = n < ((p+1) / 2) = n <= (p / 2)@. Numbers that can not be
-- uniquely mapped are called nonresidues, and there is as many residues as
-- there is nonresidues (when excluding 0). If @p = 3 `mod` 4@, the negative of
-- a residue modulo @p@ is a nonresidue and the negative of a nonresidue is a
-- residue. In other words if @n > (p / 2)@ then @p - (n² `mod` p)@
-- uniquely gives us the nonresidues. This way we have a bijection for numbers
-- @[1, p)@ instead of just @[1, p/2]@. To expand this to 128 bit values, we
-- map the remaining values to them selves.
--
-- Sources:
--
-- * <http://preshing.com/20121224/how-to-generate-a-sequence-of-unique-random-integers/ How to Generate a Sequence of Unique Random Integers>
--
-- * <https://en.wikipedia.org/wiki/Quadratic_residue Quadratic residue>

-- }}} Quadratic Residue ------------------------------------------------------

-- {{{ Bit Operations ---------------------------------------------------------

-- | Swap LO and HI halfs of a word.
--
-- @
-- 11110000 -> 00001111
-- @
--
-- Following holds:
--
-- @
-- 'swapLoAndHiHalf' '.' 'swapLoAndHiHalf' ≡ 'id'
-- @
swapLoAndHiHalf :: FiniteBits a => a -> a
swapLoAndHiHalf n = Bits.rotate n (finiteBitSize n `quot` 2)

-- | Swap every other bit in a word.
--
-- @
-- 00000000 -> 01010101
-- @
--
-- Following holds:
--
-- @
-- 'swapEveryOtherBit' '.' 'swapEveryOtherBit' ≡ 'id'
-- @
swapEveryOtherBit :: (Integral a, FiniteBits a) => a -> a
swapEveryOtherBit a = a `xor` mask
  where
    mask = foldr ((.|.) . bit) 0 [0, 2 .. finiteBitSize a]

-- }}} Bit Operations ---------------------------------------------------------
